### PR TITLE
smb: fix SMB_COM_WRITE_ANDX record parser - v2

### PR DESCRIPTION
**Pull request by b1tg**
Bug:  https://redmine.openinfosecfoundation.org/issues/6008

Previous PR: https://github.com/OISF/suricata/pull/8839

Changes from last PR:
- remove TODO & add comments

SV_BRANCH=pr/1195

### CODE REVIEW

**Comment by b1tg** [comment]
If packet has a invalid data_pad_offset, should we return an error or just let it be zero?

**Comment by b1tg** [comment]
> If packet has a invalid data_pad_offset, should we return an error or just let it be zero?

"return an error" version look like below:
```rust
    let mut data_pad_offset = data_offset.checked_sub(ax+28).ok_or(Err::Error(make_error(i, ErrorKind::LengthValue)))?;
    if high_offset.is_some() {
        data_pad_offset = data_pad_offset.checked_sub(4).ok_or(Err::Error(make_error(i, ErrorKind::LengthValue)))?;
    }
```

**Comment by catenacyber** [comment]
Do you know why there is one CI failure ?

> If packet has a invalid data_pad_offset, should we return an error or just let it be zero?

What would Windows do ?

**Comment by b1tg** [comment]
> Do you know why there is one CI failure ?

I could not reproduce this failure, and only Fedora 36 failed is strange.

> What would Windows do ?

I create a packet by use smbclient.py send a file from Linux to Win10 which have a shared folder

write_andx_request in origin packet have data_offset=0x3f

As a Rust test it looks like :

```rust
    let mut data = hex::decode("0eff000000054000000000ff00000008001400000014003f00000000001400414141414141414141414141414141414141410a").unwrap();
    let result = parse_smb1_write_andx_request_record(&data, SMB1_HEADER_SIZE);
    assert!(result.is_ok());
    let record = result.unwrap().1;
    assert_eq!(record.offset, 0);
    assert_eq!(record.len, 20);
    assert_eq!(record.fid, &[0x05, 0x40]);
    assert_eq!(record.data.len(), 20);
    assert_eq!(record.data, b"AAAAAAAAAAAAAAAAAAA\n");
```

then I use a proxy to modify data_offset  in write_andx_request before send to server:

1. change data_offset to 0x3e, 0x3d, 0x3c: Windows 10 write 20 bytes (include metadata before real data in packet)
2. change data_offset to <= 0x3b: Windows 10 return Server Error(Non specific error code)
3. change data_offset to [0x40, 0x53]: Windows 10 write [19, 0] bytes
4. change data_offset to 0x54: Windows 10 return Server Error(Non specific error code)


and there is no one byte padding in the origin packet which is a little surprise as spec [MS-CIFS].pdf says always take one byte padding

**Comment by victorjulien** [comment]
Restarted the CI job to make sure it's not a random failure.

**Comment by catenacyber** [comment]
Nice @b1tg do you have a pcap of this ?

So, I think you should return an error, but make sure that this does not prevent further parsing of SMB PDUs

Why is this PR a draft ? What are you expecting from us ?

**Comment by b1tg** [comment]
> Why is this PR a draft ? What are you expecting from us ?

victorjulien marked this pull request as draft

> do you have a pcap of this ?

Added to https://redmine.openinfosecfoundation.org/issues/6008

**Review by catenacyber — CHANGES_REQUESTED** [review]
Here is my feedback :
CI : green :-)
Code : reviewing it right now
Commits segmentation : I think the 2 commits should be squashed together
Commit messages : ok
Git ID set : looks fine for me
CLA : did you sign it ?
Doc update : not needed
Redmine ticket : ok
Rustfmt : ok, file not yet formatted
Tests : Would you be able to address the remarks in the suricata-verify PR ? (avoid the shell checks and have just json checks)
Dependencies added: none

File: rust/src/smb/smb1_records.rs [hunk 1]
```diff
@@ -106,15 +107,23 @@ pub fn parse_smb1_write_andx_request_record(i : &[u8], andx_offset: usize) -> IR
...
     //spec [MS-CIFS].pdf says always take one byte padding
-    let (i, _padding) = cond(bcc > data_len_low, |b| take(bcc - data_len_low)(b))(i)?; // TODO figure out how this works with data_len_high
-    let (i, _padding_evasion) = cond(data_offset > ax+4+2*(wct as u16), |b| take(data_offset - (ax+4+2*(wct as u16)))(b))(i)?;
+    let (i, _padding) = take(1_usize)(i)?;
+    // [MS-CIFS].pdf 2.2.4.43.1
+    // without high_offset, &SMB_Data.Bytes.Data - &SMB_Parameters = 28
+    // with high_offset, &SMB_Data.Bytes.Data - &SMB_Parameters = 32
+    let mut data_pad_offset = data_offset.saturating_sub(ax+28);
+    if high_offset.is_some() {
+        data_pad_offset = data_pad_offset.saturating_sub(4);
```
Comment (catenacyber) [inline rust/src/smb/smb1_records.rs:121]: I wonder if we can be more direct by using `wct` as was done before (`ax+4+2*(wct as u16)`)

File: rust/src/smb/smb1_records.rs [hunk 2]
```diff
@@ -862,3 +871,18 @@ pub fn parse_smb_record(i: &[u8]) -> IResult<&[u8], SmbRecord> {
     };
     Ok((i, record))
 }
+
+#[test]
+fn test_parse_smb1_write_andx_request_record() {
```
Comment (catenacyber) [inline rust/src/smb/smb1_records.rs:876]: Maybe the S-V test is enough...
This blob looks too big

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/8889

**Pull request closed without merging**